File: compatibility-tests/src/test/java/nnl/rocks/kactoos/map/BehavesAsMap.java
/**
 * The MIT License (MIT)
 * <p>
 * Copyright (c) 2017-2018 Yegor Bugayenko
 * <p>
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * <p>
 * The above copyright notice and this permission notice shall be included
 * in all copies or substantial portions of the Software.
 * <p>
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package nnl.rocks.kactoos.map;

import java.util.Map;
import org.hamcrest.Description;
import org.hamcrest.MatcherAssert;
import org.hamcrest.Matchers;
import org.hamcrest.TypeSafeMatcher;

/**
 * Matcher for collection.
 * @author Yegor Bugayenko (yegor256@gmail.com)
 * @version $Id: 872a70ef0c079221e522cb41772c36f637025563 $
 * @param <K> Type of key
 * @param <V> Type of value
 * @since 0.24
 * @checkstyle JavadocMethodCheck (500 lines)
 */
public final class BehavesAsMap<K, V> extends TypeSafeMatcher<Map<K, V>> {

    /**
     * Sample key.
     */
    private final K key;

    /**
     * Sample value.
     */
    private final V value;

    /**
     * Ctor.
     * @param akey Sample key
     * @param val Sample value
     */
    public BehavesAsMap(
        final K akey,
        final V val
    ) {
        super();
        this.key = akey;
        this.value = val;
    }

    @Override
    public boolean matchesSafely(final Map<K, V> map) {
        MatcherAssert.assertThat(map, Matchers.hasKey(this.key));
        MatcherAssert.assertThat(map, Matchers.hasValue(this.value));
        MatcherAssert.assertThat(map.keySet(), Matchers.hasItem(this.key));
        MatcherAssert.assertThat(map.values(), Matchers.hasItem(this.value));
        return true;
    }

    @Override
    public void describeTo(final Description desc) {
        desc.appendText("not a valid map");
    }
}

File: kactoos-jvm/src/test/kotlin/nnl/rocks/kactoos/map/BehavesAsMap.kt

package nnl.rocks.kactoos.map

import org.hamcrest.Description
import org.hamcrest.MatcherAssert
import org.hamcrest.Matchers
import org.hamcrest.TypeSafeMatcher

/**
 * Matcher for collection.
 * @author Yegor Bugayenko (yegor256@gmail.com)
 * @version $Id: 872a70ef0c079221e522cb41772c36f637025563 $
 * @param <K> Type of key
 * @param <V> Type of value
 * @since 0.24
 * @checkstyle JavadocMethodCheck (500 lines)
</V></K> */
class BehavesAsMap<K, V>
/**
 * Ctor.
 * @param akey Sample key
 * @param val Sample value
 */
(
        /**
         * Sample key.
         */
        private val key: K,
        /**
         * Sample value.
         */
        private val value: V
) : TypeSafeMatcher<Map<K, V>>() {

    public override fun matchesSafely(map: Map<K, V>): Boolean {
        MatcherAssert.assertThat(map, Matchers.hasKey(this.key))
        MatcherAssert.assertThat(map, Matchers.hasValue(this.value))
        MatcherAssert.assertThat(map.keys, Matchers.hasItem(this.key))
        MatcherAssert.assertThat(map.values, Matchers.hasItem(this.value))
        return true
    }

    override fun describeTo(desc: Description) {
        desc.appendText("not a valid map")
    }
}
